compilers/concrete-compiler/compiler/lib/Runtime/context.cpp

```diff
@@ -31,59 +31,148 @@ FFT::~FFT() {
+}
+
+std::pair<FFT, std::shared_ptr<std::vector<std::complex<double>>>>
+RuntimeContext::convert_to_fourier_domain(LweBootstrapKey &bsk) {
+  auto info = bsk.getInfo().asReader();
+
+  size_t decomposition_level_count = info.getParams().getLevelCount();
+  size_t decomposition_base_log = info.getParams().getBaseLog();
+  size_t glwe_dimension = info.getParams().getGlweDimension();
+  size_t polynomial_size = info.getParams().getPolynomialSize();
+  size_t input_lwe_dimension = info.getParams().getInputLweDimension();
+
+  // Create the FFT
+  FFT fft(polynomial_size);
+
+  // Allocate scratch for key conversion
+  size_t scratch_size;
+  size_t scratch_align;
+  concrete_cpu_bootstrap_key_convert_u64_to_fourier_scratch(
+      &scratch_size, &scratch_align, fft.fft);
+  auto scratch = (uint8_t *)aligned_alloc(scratch_align, scratch_size);
+
+  // Allocate the fourier_bootstrap_key
+  auto &bsk_buffer = bsk.getBuffer();
+  auto fourier_data = std::make_shared<std::vector<std::complex<double>>>();
+  fourier_data->resize(bsk_buffer.size() / 2);
+  auto bsk_data = bsk_buffer.data();
+
+  // Convert bootstrap_key to the fourier domain
+  concrete_cpu_bootstrap_key_convert_u64_to_fourier(
+      bsk_data, fourier_data->data(), decomposition_level_count,
+      decomposition_base_log, glwe_dimension, polynomial_size,
+      input_lwe_dimension, fft.fft, scratch, scratch_size);
+  free(scratch);
+
+  return std::pair<FFT, std::shared_ptr<std::vector<std::complex<double>>>>(
+      std::move(fft), fourier_data);
+}
+} // namespace concretelang
+} // namespace mlir
+
+#ifdef CONCRETELANG_DATAFLOW_EXECUTION_ENABLED
+#include "concretelang/Runtime/key_manager.hpp"
+
+// Register the HPX actions for retrieving the evaluation keys from
+// the master node (must be in global namespace)
+HPX_PLAIN_ACTION(mlir::concretelang::dfr::getKsk, _dfr_get_ksk_action)
+HPX_PLAIN_ACTION(mlir::concretelang::dfr::getBsk, _dfr_get_bsk_action)
+HPX_PLAIN_ACTION(mlir::concretelang::dfr::getPKsk, _dfr_get_pksk_action)
+
+namespace mlir {
+namespace concretelang {
+const uint64_t *DistributedRuntimeContext::keyswitch_key_buffer(size_t keyId) {
+  if (dfr::_dfr_is_root_node())
+    return RuntimeContext::keyswitch_key_buffer(keyId);
+
+  std::lock_guard<std::mutex> guard(cm_guard);
+  if (ksks.find(keyId) == ksks.end()) {
+    _dfr_get_ksk_action getKskAction;
+    dfr::KeyWrapper<LweKeyswitchKey> kskw =
+        getKskAction(hpx::find_root_locality(), keyId);
+    ksks.insert(std::pair<size_t, LweKeyswitchKey>(keyId, kskw.keys[0]));
+  }
+  auto it = ksks.find(keyId);
+  assert(it != ksks.end());
+  return it->second.getBuffer().data();
+}
+
+void DistributedRuntimeContext::getBSKonNode(size_t keyId) {
+  assert(fbks.find(keyId) == fbks.end());
+  assert(dffts.find(keyId) == dffts.end());
+  _dfr_get_bsk_action getBskAction;
+  dfr::KeyWrapper<LweBootstrapKey> bskw =
+      getBskAction(hpx::find_root_locality(), keyId);
+
+  auto fdbsk = convert_to_fourier_domain(bskw.keys[0]);
```

Comment:
If I well understand we always redo the conversion to the fourier domain when fetching bsk keys on node, while on the root node we do the conversion for all bsk keys. I'm not sure it's voluntary to do that, maybe the idea is to not wait that the root node have done all the conversion before starting to compute on slave node?
Anyway I think we could try to fetch the fourrier one if already converted on the root node, also the root node don't need to convert all the bsk.
I feel it is not mandatory for this PR at all and it is maybe micro optimization, but we could discuss of that, say me wdyt.
I'm also wondering if we explicitly modelize this buffer management of the keys in MLIR it could be easier and more generic to solve this kind of problem?

Reply:
I fully agree - this is just providing the functionality for now, but we will need to look at how we can optimise. It's difficult to be very flexible with the ordering and say we want to start work on compute nodes early as the work is created on the master side. This kind of targeted optimisations are natural for code written directly for distributed systems, but a lot harder for code generated from a sequential source.